eip-0008.md

```diff
@@ -0,0 +1,425 @@
+
+## Background And Motivation
+
+To prevent DoS attacks every script in the Ergo blockchain (represented as
+[ErgoTree](https://ergoplatform.org/docs/ErgoTree.pdf)) has to be executed with the
+control of the execution costs (aka _costing_). This is necessary to prevent delayed
+validations of new block candidates by the network nodes. Both over-sized and
+over-expensive scripts may be exploited to attack the network by exhausting verifiers'
+resources.
+
+There are two approaches to do cost control: 
+1) Ahead-Of-Time costing (AOTC) - an estimation of the execution costs, where the cost of
+the script is approximated before and without the script execution.
+2) Just-In-Time costing (JITC) - cost accumulation during contract execution limited by
+some value. (aka Gas limit in Ethereum)
+
+Please see [Appendix A](#appendix-a-aot-vs-jit-costing) for AOTC vs JITC comparison. The
+potential benefits of JITC are described in [Appendix
+B](#appendix-b-benefits-of-jit-costing).
+
+## Ergo Transaction Validation With JIT Costing
+
+This EIP describes a new version of transaction validation, which uses JIT costing (JITC)
+for script execution. JITC which allows to prevent execution of scripts beyond allotted
+limits.
+
+The JITC implementation doesn't require changes in the serialization format and
+interpretation of transactions in the block. Only cost verification is changed with this
+proposal, which however require activation via hard-fork. The current version of Ergo
+Protocol v1 will be increased to v2.
+
+The proposed change is conservative so that the new Ergo Protocol v2 is backward
+compatible:
+1) All existing applications (like exchanges, ErgoMix, DEX) will not require any changes.
+2) The new protocol doesn't incur additional overhead and instead reduce it. 
+
+The cost calculation of ErgoTrees is implemented as it is defined in [the next
+section](#new-costing-of-ergotree).
+
+We first describe how transactions of a block are validated. This _on-chain_ validation is
+part of the consensus protocol. Transaction validation is a deterministic procedure which
+should be the same for all nodes of the network.
+
+Next, we describe the steps which should be done _off-chain_ by applications in order to
+create, sign and send new transactions.
+
+### Transaction Verification
+
+The verification is performed by _verifier_ nodes for each new transaction `tx` either
+selected from the pool or contained in a new block candidate broadcast over the network.
+
+1) A verifier determines the `costLimit` parameter of the transaction which should be less
+than `MaxTxCost`. The limit is computed as `costLimit = tx.feeOut.value /
+tx.feeOut.R4[Long]` if the register is not empty, otherwise `costLimit = MaxTxCost`. 
+The value in the `R4` register is the NERG price of a unit of cost (`costPrice`).
+The `MaxTxCost` parameter is described in the [section below](#maximum-transaction-cost).
```

Comment:
Giving such semantics to R4 already breaks backwards compatibility for all the apps and standards using registers (mixer, EIP-4, DEX etc)